### PR TITLE
Grant IAM rights to service identities in host project

### PR DESCRIPTION
**Pull request by wiktorn**
Allow granting IAM permissions in Shared VPC host project based on list of services to be used in service project.
* [x] leverage this functionality in `project-factory` blueprint

Closes: #1527, #1526, #1518

### CODE REVIEW

File: modules/project/shared-vpc.tf [hunk 1]
```diff
@@ -17,15 +17,25 @@
 # tfdoc:file:description Shared VPC project-level configuration.
 
 locals {
+  _shared_vpc_agent_config = yamldecode(file("${path.module}/sharedvpc-agent-iam.yaml"))
+  _shared_vpc_agent_config_filtered = [
+    for config in local._shared_vpc_agent_config : config
+    if contains(var.shared_vpc_service_config.grant_iam_for_services, config.service)
```
Comment (juliocc) [inline modules/project/shared-vpc.tf:23]: I think you should also filter this with the apis that are enabled in the project. In other words, something like `contains(stintersection(var.services, var.shared_vpc_service_config.grant_iam_for_services), config.service)`

File: modules/project/shared-vpc.tf [hunk 2]
```diff
@@ -34,11 +44,9 @@ locals {
       try(var.shared_vpc_host_config.service_projects, null), []
     )
   }
-  svpc_service_config = coalesce(var.shared_vpc_service_config, {
-    host_project = null, service_identity_iam = {}
-  })
+
   svpc_service_iam = {
-    for b in local._svpc_service_iam : "${b.role}:${b.service}" => b
+    for b in distinct(concat(local._svpc_service_iam, local._shared_vpc_agent_grants)) : "${b.role}:${b.service}" => b
```
Comment (juliocc) [inline modules/project/shared-vpc.tf:49]: `distinct+concat`, or `setunion`

**Review by juliocc — COMMENTED** [review]
I like it

**Review by juliocc — APPROVED** [review]

File: modules/project/shared-vpc.tf [hunk 3]
```diff
@@ -59,7 +67,7 @@ resource "google_compute_shared_vpc_service_project" "service_projects" {
 
 resource "google_compute_shared_vpc_service_project" "shared_vpc_service" {
   provider        = google-beta
-  count           = local.svpc_service_config.host_project != null ? 1 : 0
+  count           = var.shared_vpc_service_config.host_project != null ? 1 : 0
```
Comment (juliocc) [inline modules/project/shared-vpc.tf:70]: One more thing. Can you add a precondition here to ensure that `grant_iam_for_services` only contains APIs mentioned in the yaml?

File: modules/project/shared-vpc.tf (hunk 1 above)
Comment (ludoo) [reply modules/project/shared-vpc.tf:23]: we might not have the APIs here Julio

File: modules/project/variables.tf [hunk 4]
```diff
@@ -292,12 +292,24 @@ variable "shared_vpc_host_config" {
...
 variable "shared_vpc_service_config" {
   description = "Configures this project as a Shared VPC service project (mutually exclusive with shared_vpc_host_config)."
-  # the list of valid service identities is in service-accounts.tf
+  # the list of valid service identities is in service-agents.yaml
   type = object({
-    host_project         = string
-    service_identity_iam = optional(map(list(string)))
+    host_project           = string
+    service_identity_iam   = optional(map(list(string)), {})
+    grant_iam_for_services = optional(list(string), [])
```
Comment (ludoo) [inline modules/project/variables.tf:299]: can we rename to `service_identity_iam_grants`?

File: modules/project/README.md [hunk 5]
```diff
@@ -252,6 +252,29 @@ module "service-project" {
...
+  }
+}
+
+module "service-project" {
+  source = "./fabric/modules/project"
+  name   = "my-service-project"
+  shared_vpc_service_config = {
+    host_project = module.host-project.project_id
+    grant_iam_for_services = [
+      "container.googleapis.com",
```
Comment (ludoo) [inline modules/project/README.md:271]: I would skip `.googleapis.com`.

**Review by ludoo — APPROVED** [review]
Very nice, dropped a couple comments but I really like this!

File: modules/project/README.md (hunk 5 above)
Comment (juliocc) [reply modules/project/README.md:271]: I kind of like the googleapis suffix as it kind of conveys that the field only makes sense if you're using that API (i.e. you wouldn't put container here if you don't enable the container API)

Comment (ludoo) [reply modules/project/README.md:271]: I see your point, it's fine it's not like typing it 2-3 times would be a problem

File: modules/project/variables.tf (hunk 4 above)
Comment (wiktorn) [reply modules/project/variables.tf:299]: Sound good until I realize that I have `service_identity_iam` just above. Maybe `services_iam_grants`?

File: modules/project/shared-vpc.tf (hunk 1 above)
Comment (wiktorn) [reply modules/project/shared-vpc.tf:23]: We can make it more complex like:
```hcl
    if contains(var.shared_vpc_service_config.services_iam_grants, config.service) && (
      var.services == null || contains(var.services, config.service)
    )
```

If you add more services here, than you activate, the should fail on missing service identity. Filtering here, will make sure that there is no failure at the cost of not letting the user now, that his intent was not fulfilled. 

So I prefer not to add such filtering.

File: modules/project/variables.tf (hunk 4 above)
Comment (ludoo) [reply modules/project/variables.tf:299]: `service_iam_grants` ? :)

File: modules/project/shared-vpc.tf (hunk 3 above)
Comment (wiktorn) [reply modules/project/shared-vpc.tf:70]: One of the usage patterns that I would like to allow is something like that:
```hcl
module "service-project" {
  source          = "./fabric/modules/project"
  services = [
    "alloydb.googleapis.com",
    "run.googleapis.com",
  ]
  shared_vpc_service_config = {
    host_project = module.project.project_id
    grant_iam_for_services = module.service-project.services
  }
}
```

I know that I'm contradicting here a bit myself from the above comment, but ability not to repeat myself, and not having to check in YAML file, which services are allowed here, is reducing overhead when using this module.

Let's allow user to have magic in their own code 🧙

File: modules/project/shared-vpc.tf (hunk 2 above)
Comment (wiktorn) [reply modules/project/shared-vpc.tf]: Thanks, corrected.

File: modules/project/variables.tf (hunk 4 above)
Comment (wiktorn) [reply modules/project/variables.tf]: Thanks

File: blueprints/factories/project-factory/main.tf [hunk 6]
```diff
@@ -131,19 +144,18 @@ locals {
     coalesce(var.labels, {}), coalesce(try(var.defaults.labels, {}), {})
   )
   services = distinct(concat(var.services, local._services))
-  vpc = coalesce(var.vpc, {
-    host_project = null, gke_setup = null, subnets_iam = null
-  })
   vpc_cloudservices = (
-    local.vpc_gke_service_agent ||
+    var.vpc.gke_setup.enable_host_service_agent ||
     contains(var.services, "compute.googleapis.com")
```
Comment (wiktorn) [inline blueprints/factories/project-factory/main.tf:149]: Anybody can give a reference to docs why do we grant `roles/compute.networkUser` to cloudservices if using Compute Engine? 

Should this be added to `sharedvpc-agent-iam.yaml`?

Comment (ludoo) [reply blueprints/factories/project-factory/main.tf:149]: MIGs managing autoscaling VMs

**Pull request merged**